docs: Add libFuzzer Implementation Documents (#38)
fuzzuf/fuzzuf: docs/algorithms/libfuzzer/implementation_en.md

# libFuzzer Implementation in fuzzuf

This section describes the implementation of libFuzzer in fuzzuf.

## How to Construct the Standard libFuzzer

The fuzzer corresponding to `Fuzzer::Loop()` of the original libFuzzer is `createRunone()` in [include/fuzzuf/algorithms/libfuzzer/create.hpp](/include/fuzzuf/algorithms/libfuzzer/create.hpp). You can use the libFuzzer implementation in fuzzuf by using this function when you use it as a standard libFuzzer.

## HierarFlow Nodes for libFuzzer Implementation

> We implement the following nodes to represent libFuzzer in HierarFlow. For more details about each node, refer to the comments in the source code or the documentation generated by Doxygen.

Reviewer: ```suggestion
fuzzuf has the following nodes to represent libFuzzer in HierarFlow. For more details about each node, refer to the comments in the source code or the documentation generated by Doxygen.
```
This sounds more natural, personally.

This mutator is unimplemented in fuzzuf because it does not support `trace-cmp` to get information about comparison operations.

#### Mutate_AddWordFromPersistentAutoDictionary

libFuzzer has a dictionary that stores the values inserted by `Mutate_AddWordFromManualDictionary` and `Mutate_AddWordFromTORC` that led to the discovery of new coverage. Since CMP is unimplemented in fuzzuf, this mutator only stores input from `Mutate_AddWordFromManualDictionary`.

#### Custom Mutator

libFuzzer provides CustomMutator and CustomCrossOver to add your mutators. fuzzuf does not provide nodes for these mutators because it is easier to add nodes by yourself than provide special nodes.

### Corpus

libFuzzer can persist the complete state of corpus and resume fuzzing. Still, the implementation of fuzzuf only persists the input values, so it cannot completely restore the previous state even if it resumes from the persistent information.

### Feature

#### Data Flow Trace

> Data Flow Trace uses a record of data movement obtained using LLVM's DataFlowSanitizer to determine which parts of the input values affect the branch. By masking the range of mutation based on this result, we can concentrate on finding the input that exits a particular branch. However, the original implementation of libFuzzer does not use this information to generate the mask.

Reviewer: ```suggestion
Data Flow Trace uses a record of data movement obtained using LLVM's DataFlowSanitizer to determine which parts of the input values affect the branch. By masking the range of mutation based on this result, a fuzzer can concentrate on finding the input that exits a particular branch. However, the original implementation of libFuzzer does not use this information effectively to generate the mask.
```

This feature is not implemented in fuzzuf because there is no way to get the information equivalent to DataFlowSanitizer similar to CMP.

### Executor

#### Avoiding Child Process Creation

libFuzzer avoids the cost of creating child processes by linking the fuzzing target and the fuzzer into the same binary. Still, since there is no equivalent executor in fuzzuf, the implementation of fuzzuf creates child processes.

#### Support for Shared Libraries

> libFuzzer has a mechanism to collect and combine the edge coverage of executable binaries and the edge coverage of shared libraries linked to them. This feature is not implemented in fuzzuf because there is no way to get coverage from shared libraries.

Reviewer: ```suggestion
libFuzzer has a mechanism to collect and combine the edge coverage of executable binaries and that of shared libraries linked to them. This feature is not implemented in fuzzuf because there is no way to get coverage from shared libraries.
```

Author: ```suggestion
libFuzzer has a mechanism to collect and combine the edge coverage of both executable binaries and shared libraries linked to them. This feature is not implemented in fuzzuf because there is no way to get coverage from shared libraries.
```
To avoid the duplicated "the edge coverage" words, how about this fix?
